README.md
### Purpose

The purpose of Glyco is to provide a platform for tracking diabetes, along with serving as a user friendly service to consolidate, analyze, and improve the user's wellbeing. The application has features including tracking glucose levels, A1C levels, food intake, exercise, and step amounts. Glyco provides an innovative and efficient way of inputting these measurements - shortcuts provide ease of access and encourage its users to track their measurements by cutting down the time it takes to input them, a quality that distinguishes it from other diabetes-tracking applications. It also acts as an integration point by grabbing data from applications that the patient already uses to consolidate all of that information into one place (e.g. incorporating glucose levels and steps from Apple Health). 

By analyzing past data, Glyco also encourages better health habits by generating health challenges for the patient to improve their condition and provides intuitive data analysis so the patient can track their progress over both week and month time periods. Glyco serves people who currently have Type I and Type II diabetes, people with prediabetes, and their caretakers if they're not capable of using the application. More specifically, it aims to serve people of all ages - from children to young adults to older adults - as the measurements that people with diabetes have to be aware of tend to be similar regardless of age or type of diabetes.

With time-efficient measurement tracking, data consolidation and visualization, and health challenge generation, Glyco serves as an application that not only provides ease to their users but also attempts to improve their wellbeing.

### Dependencies

More information about versions can be found in the pubspec.yaml file.

* Flutter
* Font Awesome for Flutter
* Nano Healthkit
* fl_chart
* Cupertino Icons
* Mockito

### Deployment

For build and deployment, our team relied on Flutter's automated building and fastlane. For more detailed instruction, visit: https://medium.com/@pavel.sulimau/deliver-flutter-ios-app-to-testflight-with-fastlane-1513ae047cd9.

1. Ensure that you have an Apple Developer account
2. Install fastlane on your device. "brew install fastlane" for macOS, or visit https://docs.fastlane.tools/getting-started/ios/setup/ for instructions on installing fastlane
3. Ensure that the bundle ID written in the /ios/Runner.xcodeproj/project.pbxproj file is correct
4. Open /ios/Runner.xcodeproj in Xcode. Ensure that the correct team and bundle ID are displayed, and that iOS deployment target is set to 12.0
5. "cd glyco/" and run the command "flutter build ios --release --no-codesign"
6. "cd ios/" and run the command "bundle exec fastlane beta". The application should then be placed on AppStoreConnect

RELEASE_NOTES.txt
Glyco Release Notes

Current Release Number: 1.0
Release Date: 11/16

Note: Though the Flutter framework allows for developing for both iOS and Android apps, this application was exclusively designed for and tested on the iPhone 11. This application also requires that the iOS deployment versions have to exclusively be in the range between 9.0 to 14.1.99.

Features: 

1) Users can create accounts and sign into existing accounts. These features are supported by the 'Forgot Password' feature, which sends the users an email to resolve this issue.

2) Glyco integrates with and pulls data from the Apple Health app. Users allow permissions upon creating an account so that Glyco can pull values of glucose levels and/or steps from Apple Health. These values are pulled into the application and updated on the database and, subsequently, the user's measurements.

3) Measurement tracking - The application tracks singular values for glucose levels, activity time, steps, carb intake, and A1C levels for each day. A calendar feature allows the user to switch back to previous days in order to view past measurements or add measurements to past days. Users may edit these values by clicking on any of the measurement tiles except steps amount and directly editing them.

4) The shortcuts feature, which distinguishes the user experience of Glyco from other applications. There are icons representing specific values that are either set by default or edited by the user to reflect their usual lifestyle. By clicking on these icons, their daily measurements are automatically updated. A banner that displays immediately after clicking the shortcuts will provide the option to undo an added measurement. Shortcuts can be edited in the Settings screen.

5) Glyco analyzes the user's past data and generates challenges that encourages the user to improve their health habits, either through studying steps, activity time, or carb intake. For example, if a user has been walking less on average this week than they have been during the rest of the month, Glyco will generate a challenge encouraging the user to increase the current average step amount to 25% higher than it currently is.

6) Glyco's challenge generation is supplemented by the progress feature that lives on the Challenges screen. Glyco will track the user's progress towards that goal and visualize a pie chart with that progress. This progress feature also includes a weekly chart visualizing how the user's glucose levels for the last 7 days. Both of these charts are interactive - holding onto the glucose chart will display the date and glucose level of that day, while holding onto the pie chart sections will animate them to grow in size.

7) Glyco has features to edit the user's account settings. Users are allowed to change their name, email address, and passsword after creating their account. As mentioned earlier, they can also edit their shortcut values to better reflect their lifestyle.

8) Glyco's 'View Monthly Data' feature allows user's to see Glyco's analysis and visualization of their past data. This feature takes the user's past 30 days worth of data and visualizes several charts. These include a pie chart displaying the range and percentage of glucose measurements in relation to the daily recommended glucose levels, a line chart displaying individual glucose measurements every day, a bar chart displaying individual exercise times every day, and a pie chart displaying the range and percentage of carb intake measurements in relation to the daily recommended carb intake.

Known Bugs or Limitations:
- Glyco is currently limited because it only takes singular values for each of the measurements (glucose levels, steps, activity time, carb intake, and A1C levels). This reduces the accuracy of our analysis - for example, glucose levels are analyzed by the most recently updated value, rather than their average value for that day. This is elaborated on in BACKLOG.txt.
- Glyco fetches to the database every time the user navigates to the Measures screen. This is inefficient because they are unnecessary calls happening, while standard practice would be to call once at the opening of the app and using the data collected then. (There could then be a refresh button that intentionally fetches from the database if needed.)
- Currently, Glyco asks the user to refresh any time they navigate to a day that has not yet been created for that user. This isn't user friendly and should be addressed by having the application automatically refresh itself after making a call to the database to create an additional date.
- The excercise bar chart in 'View Monthly Data' currently shows the value '1.0' for all '0' values

glyco/lib/providers/challenges.dart
import 'package:http/http.dart' as http;
import 'package:flutter/material.dart';
import 'dart:convert';
import './measurement.dart';
import './measurements.dart';

class Challenges with ChangeNotifier {
  String _providerChallengeGiven = 'null';
  int _providerChallengeGoal = -1;

  // Measurements measurements = ;

  List<Measurement> _measurements = [];

  List<Measurement> get measurements {
    return [..._measurements];
  }

  Measurement get latestMeasurement {
    return _measurements.last;
  }

  Measurement findByDate(DateTime date) {
    return _measurements.firstWhere(
      (measurement) =>
          measurement.date.year == date.year &&
          measurement.date.month == date.month &&
          measurement.date.day == date.day,
      orElse: () {
        // TODO: handle this
        print('no element');
        Measurement newMeasurement = Measurement(
          id: DateTime.now().toIso8601String(),
          a1cLevel: 0,
          avgGlucoseLevel: 0,
          calories: 0,
          carbs: 0,
          currGlucoseLevel: 0,
          exerciseTime: 0,
          date: date,
          steps: 0,
          lastUpdate: date,
        );
        addMeasurement(newMeasurement).then((value) => print("added"));
      },
    );
  }

  int roundToMultiple(double number, int multiple) {
    if (number == 0) {
      return multiple;
    }

    int result = multiple;

    if (number % multiple == 0) {
      return number.round();
    } else {
      int division = ((number / multiple) + 1).round();
      result = division * multiple;
    }
    return result;
  }

  // STEPS AVERAGES

  double monthSteps() {
    int totalSteps = 0;
    int numDays = 0;

    for (int i = 7; i < 31; i++) {
      DateTime day = DateTime.now().subtract(Duration(days: i));

      totalSteps += findByDate(day).steps;
      numDays++;
    }

    double avgSteps = (totalSteps / numDays);
    return avgSteps;

    // return 9780;
  }

  double weekSteps() {
    int totalSteps = 0;
    int numDays = 0;

    for (int i = 1; i < 7; i++) {
      DateTime day = DateTime.now().subtract(Duration(days: i));

      totalSteps += findByDate(day).steps;
      numDays++;
    }

    double avgSteps = (totalSteps / numDays);
    return avgSteps;

    // return 2590;
  }

  // ACTIVITY AVERAGES
  double monthActivity() {
    int totalActivity = 0;
    int numDays = 0;

    for (int i = 7; i < 31; i++) {
      DateTime day = DateTime.now().subtract(Duration(days: i));

      totalActivity += findByDate(day).exerciseTime;
      numDays++;
    }

    double avgActivity = (totalActivity / numDays);
    return avgActivity;

    // return 30;
  }

  double weekActivity() {
    int totalActivity = 0;
    int numDays = 0;

    for (int i = 1; i < 7; i++) {
      DateTime day = DateTime.now().subtract(Duration(days: i));

      totalActivity += findByDate(day).exerciseTime;
      numDays++;
    }

    double avgActivity = (totalActivity / numDays);
    return avgActivity;

    // return 28;
  }

  // CARBS AVERAGES
  double monthCarbs() {
    int totalCarbs = 0;
    int numDays = 0;

    for (int i = 7; i < 31; i++) {
      DateTime day = DateTime.now().subtract(Duration(days: i));

      totalCarbs += findByDate(day).carbs;
      numDays++;
    }

    double avgCarbs = (totalCarbs / numDays);
    return avgCarbs;

    // return 10;
  }

  double weekCarbs() {
    int totalCarbs = 0;
    int numDays = 0;

    for (int i = 1; i < 7; i++) {
      DateTime day = DateTime.now().subtract(Duration(days: i));

      totalCarbs += findByDate(day).carbs;
      numDays++;
    }

    double avgCarbs = (totalCarbs / numDays);
    return avgCarbs;

    // return 60;
  }

  String generateChallenge() {
    double stepsReduction = 0;
    int stepsReductionPercent = 0;
    double activityReduction = 0;
    int activityReductionPercent = 0;
    double carbDifference = 0;

    if (monthSteps() != 0) {
      stepsReduction = 1 - (weekSteps() / monthSteps());
      stepsReductionPercent = (stepsReduction * 100).round();
    }
    if (monthActivity() != 0) {
      activityReduction = 1 - (weekActivity() / monthActivity());
      activityReductionPercent = (activityReduction * 100).round();
    }
    if (monthCarbs() != 0) {
      carbDifference = (weekCarbs() / monthCarbs());
    }

    // STEPS

    // If their weekly average has increased compared to the rest of the month
    if (weekSteps() > monthSteps()) {
      int stepsGoal = roundToMultiple(weekSteps(), 100);
      _providerChallengeGiven = 'steps';
      _providerChallengeGoal = stepsGoal;

      if (monthSteps() == 0) {
        return 'Your steps this week have been much higher than the rest of the month. Keep up the good work!';
      }

      int stepsIncreasePercent =
          (((weekSteps() / monthSteps()) - 1) * 100).round();
      return 'Your steps this week have been ' +
          stepsIncreasePercent.toString() +
          '% higher than the rest of the month. Keep up the good work!';
    }
    // If their weekly average has decreased more than 85%, generate a challenge that is 25% higher than their weekly average
    if (stepsReduction >= 0.15) {
      int stepsGoal = roundToMultiple((weekSteps() * 1.25), 100);
      _providerChallengeGiven = 'steps';
      _providerChallengeGoal = stepsGoal;

      return 'Your steps for the last week have been ' +
          stepsReductionPercent.toString() +
          "% lower than the rest of the month. Try to get " +
          stepsGoal.toString() +
          " steps this week!";
    }
    // If their weekly average is less than the recommended daily amount of steps
    if (weekSteps() < 3000) {
      _providerChallengeGiven = 'steps';
      _providerChallengeGoal = 4000;
      return 'Your steps are below the recommended daily steps. Try to get to 4,000 steps this week! ';
    }
    // If their weekly average is greater than the recommended daily amount of steps
    if (weekSteps() >= 10000) {
      _providerChallengeGiven = 'steps';
      _providerChallengeGoal = roundToMultiple(weekSteps(), 100);
      return 'Congratulations! You have hit the daily recommended step intake of 10000 steps. Keep up the good work!';
    }

    // ACTIVITY
    if (weekActivity() > monthActivity()) {
      int activityGoal = roundToMultiple(weekActivity(), 10);
      _providerChallengeGiven = 'activity';
      _providerChallengeGoal = activityGoal;

      if (monthActivity() == 0) {
        return 'Your activity time this week has been higher than the rest of the month. Keep up the good work!';
      }

      int activityIncreasePercent =
          (((weekActivity() / monthActivity()) - 1) * 100).round();
      return 'Your activity time this week has been ' +
          activityIncreasePercent.toString() +
          '% higher than the rest of the month. Keep up the good work!';
    }
    if (activityReduction >= 0.15) {
      int activityGoal = roundToMultiple((weekActivity() * 1.25), 10);
      _providerChallengeGiven = 'activity';
      _providerChallengeGoal = activityGoal;

      return 'Your activity time for the last week has been ' +
          activityReductionPercent.toString() +
          "% lower than the rest of the month. Try to get " +
          activityGoal.toString() +
          " minutes a day this week!";
    }
    if (weekActivity() < 30) {
      _providerChallengeGiven = 'activity';
      _providerChallengeGoal = 30;
      return 'Your activity time is below the recommended daily activity. Try to get to 30 minutes a day this week! ';
    }
    if (weekActivity() >= 30) {
      _providerChallengeGiven = 'activity';
      _providerChallengeGoal = roundToMultiple(weekActivity(), 10);
      return 'Congratulations! You have hit the daily recommended activity level of 30 minutes. Keep up the good work!';
    }

    // CARBS
    if (weekCarbs() < monthCarbs()) {
      int carbGoal = roundToMultiple(weekCarbs(), 10);
      _providerChallengeGiven = 'carbs';
      _providerChallengeGoal = carbGoal;

      int carbDecreasePercent =
          ((1 - (weekCarbs() / monthCarbs())) * 100).round();
      return 'Your carb intake this week has been ' +
          carbDecreasePercent.toString() +
          '% lower than the rest of the month. Keep up the good work!';
    }
    if (carbDifference > 1) {
      double carbIncrease = carbDifference - 1;
      int carbIncreasePercent = ((carbDifference - 1) * 100).round();

      if (carbIncrease >= 0.15) {
        int carbDecreaseGoal = (weekCarbs() * 0.85).round();
        _providerChallengeGiven = 'carbs';
        _providerChallengeGoal = carbDecreaseGoal;
        return 'Your carb intake for the last week has been ' +
            carbIncreasePercent.toString() +
            "% higher than the rest of the month. Try to get " +
            carbDecreaseGoal.toString() +
            " grams of carbs this week!";
      }
    }
    if (weekCarbs() > 75) {
      _providerChallengeGiven = 'carbs';
      _providerChallengeGoal = 100;
      return 'Your carb intake is above the recommended daily carb intake. Try to get down to 100g of carbs this week! ';
    }
    if (weekCarbs() <= 45) {
      _providerChallengeGiven = 'carbs';
      _providerChallengeGoal = roundToMultiple(weekCarbs(), 10);
      return 'Congratulations! You are around the daily recommended carb intake of 40g of carbs. Keep up the good work!';
    }
    if (monthCarbs() == 0 && weekCarbs() > 0) {
      _providerChallengeGiven = 'carbs';
      int carbDecreaseGoal = (weekCarbs() * 0.85).round();
      _providerChallengeGoal = carbDecreaseGoal;

      return 'Your carb intake for the last week has been higher than the rest of the month. Try to get ' +
          carbDecreaseGoal.toString() +
          " grams of carbs this week!";
    }

    _providerChallengeGiven = 'none';
    return 'There are currently no challenges. Come back later!';
  }

  String getChallenge() {
    return _providerChallengeGiven;
  }

  int getChallengeGoal() {
    return _providerChallengeGoal;
  }

  Image getProgressAsset() {
    if (_providerChallengeGiven == 'steps') {
      return Image.asset('assets/images/challenges_image1.jpg');
    }
    if (_providerChallengeGiven == 'activity') {
      return Image.asset('assets/images/challenges_image2.jpg');
    }
    if (_providerChallengeGiven == 'carbs') {
      return Image.asset('assets/images/challenges_image3.jpg');
    }
    return Image.asset('assets/images/challenges_image1.jpg');
  }

  String progressUpdate(
      Measurement measurement, String challengeGiven, int challengeGoal) {
    if (challengeGiven == 'steps') {
      int stepsProgress = measurement.steps;
      int stepsProgressPercent =
          ((stepsProgress / challengeGoal) * 100).round();
      return 'You have completed ' +
          stepsProgressPercent.toString() +
          '% of your goal, with ' +
          stepsProgress.toString() +
          ' out of ' +
          challengeGoal.toString() +
          ' steps.';
    }
    if (challengeGiven == 'activity') {
      int activityProgress = measurement.exerciseTime;
      int activityProgressPercent =
          ((activityProgress / challengeGoal) * 100).round();
      return 'You have completed ' +
          activityProgressPercent.toString() +
          '% of your goal, with ' +
          activityProgress.toString() +
          ' out of ' +
          challengeGoal.toString() +
          ' minutes.';
    }
    if (challengeGiven == 'carbs') {
      int carbProgress = measurement.carbs;
      int carbProgressPercent = ((carbProgress / challengeGoal) * 100).round();
      return 'You have completed ' +
          carbProgressPercent.toString() +
          '% of your goal, with ' +
          carbProgress.toString() +
          ' out of ' +
          challengeGoal.toString() +
          ' grams of carbs.';
    }

    return 'You currently have no progress because you have no challenges.';
  }

  Future<void> fetchAndSetMeasurements() async {
    const url = 'https://glyco-6f403.firebaseio.com/measurements.json';
    try {
      final response = await http.get(url);
      final extractedData = json.decode(response.body) as Map<String, dynamic>;
      if (extractedData == null) {
        return;
      }
      final List<Measurement> loadedMeasurements = [];
      extractedData.forEach((measurementId, measurementData) {
        loadedMeasurements.add(
          Measurement(
            id: measurementId,
            a1cLevel: measurementData['a1cLevel'],
            avgGlucoseLevel: measurementData['avgGlucoseLevel'],
            calories: measurementData['calories'],
            carbs: measurementData['carbs'],
            currGlucoseLevel: measurementData['currGlucoseLevel'],
            exerciseTime: measurementData['exerciseTime'],
            date: DateTime.parse(measurementData['date']),
            steps: measurementData['steps'],
            lastUpdate: DateTime.parse(measurementData['lastUpdate']),
          ),
        );
      });
      _measurements = loadedMeasurements;
      notifyListeners();
    } catch (error) {
      throw error;
    }
  }

  Future<void> addMeasurement(Measurement measurement) async {
    const url = 'https://glyco-6f403.firebaseio.com/measurements.json';
    try {
      final response = await http.post(url,
          body: json.encode({
            'a1cLevel': measurement.a1cLevel,
            'avgGlucoseLevel': measurement.avgGlucoseLevel,
            'calories': measurement.calories,
            'carbs': measurement.carbs,
            'currGlucoseLevel': measurement.currGlucoseLevel,
            'exerciseTime': measurement.exerciseTime,
            'date': measurement.date.toIso8601String(),
            'steps': measurement.steps,
            'lastUpdate': measurement.lastUpdate.toIso8601String(),
          }));
      final newMeasurement = Measurement(
        id: json.decode(response.body)['name'],
        a1cLevel: measurement.a1cLevel,
        avgGlucoseLevel: measurement.avgGlucoseLevel,
        calories: measurement.calories,
        carbs: measurement.carbs,
        currGlucoseLevel: measurement.currGlucoseLevel,
        exerciseTime: measurement.exerciseTime,
        date: measurement.date,
        steps: measurement.steps,
        lastUpdate: measurement.lastUpdate,
      );
      _measurements.add(newMeasurement);
      notifyListeners();
    } catch (error) {
      throw error;
    }
  }
}
